chore(v2): replace few Lodash methods with native counterparts

Simek opened the pull request:
## Motivation

Same motivation like in previous PRs.

This is the initial step in getting rid of main `lodash` package in core v2 Docusaurus package. In this PR `_.compact` and `_.flatten` methods has been replaced with their native counterparts.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Same test results before and after code refactor.

## Related PRs

* #2460
* #2519

yangshun commented:
I think I'd stick with the Lodash versions given they are more readable. Sorry about that.

Simek commented:
@yangshun I think that it's not a good approach. This means that v2 core will have two more packages (`lodash.concat` and `lodash.flatten`) and two more dev packages (`@types/lodash.concat` and  `@types/lodash.flatten`) just because of personal preferences.

yangshun commented:
@lex111 what do you think? Do you agree?

Simek commented:
Also `compact` + `flatten` replacement could be extracted to the separate util/helper method if you think this improve readability, but I have no idea where this kind of code should be placed (in which package, which path).

Simek commented:
I have updated the reworked PR and removed also `assign` and `isArray` methods. All commits has been squashed. Current version can be seen here: https://github.com/Simek/docusaurus/commit/2084e9981cf682772bd63604a6478c0aee6a0c6f

After applying those changes only `has`, `isPlainObject` and `isString` packages are needed in the core. I have a finishing commit stashed, which removes `lodash` dependency and adds listed above single method packages and typing for them. 

Let me know if I should commit those changes in the separate PR after this part will be resolved or add them in reworked version of this PR.

yangshun commented:
Ok let's do this. Could you reopen a new PR? I can't seem to reopen the PR now because it has been force pushed

slorber commented:
Hey, was wondering what is the motivation to use lodash es exactly?

Personnally I find it highly annoying to have to update package.json everywhere I need a new lodash method.

What's the problem with a regular lodash available in node?

I understand we want to optimize client code, but don't really see the value for node code.

Btw, nothing will enforce the rule of "only lodash ES imports" over time. In fact, some code of mine was already merged and does not use lodash es but regular lodash, so wonder what to do about this?

If there are striong reasons to use lodash ES then we should enforce it with an ESLint import rule, but honnestly I'm not convinced by the value of enforcing such rule (at least for node code)

slorber commented:
What I'd suggest is:
- prevent lodash import in client code by a nested .eslintrc in client code folders, with import rules preventing import from `lodash`
- allow using `lodash` in any server code, not necessarily from ES imports like this refactor introduced
